dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,190 @@
+
+    The Snowflake connection and Synapse connection IDs, as well as the time window for recent downloads,
+    are provided via DAG parameters.
+    """
+
+    @task
+    def get_trending_project_snapshot(**context) -> List[SnapshotMetrics]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'project'
+                    AND IS_PUBLIC = TRUE
+                ),
+                RECENT_DOWNLOADS AS (
+                    SELECT PROJECT_ID, RECORD_DATE, USER_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILEDOWNLOAD
+                    WHERE DATE_TRUNC('MONTH', RECORD_DATE) = DATE_TRUNC('MONTH', DATE({context["params"]["month_to_run"]}))
+                    AND STACK = 'prod'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                ),
+                TOP_10_PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID, COUNT(DISTINCT USER_ID) AS N_UNIQUE_USERS
+                    FROM RECENT_DOWNLOADS
+                    GROUP BY PROJECT_ID
+                    ORDER BY N_UNIQUE_USERS DESC
+                    LIMIT 10
+                ),
+                LATEST_FILE_HANDLES AS (
+                    SELECT ID, PROJECT_ID, FILE_HANDLE_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'file'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM TOP_10_PUBLIC_PROJECTS)
+                ),
+                FILE_SIZES AS (
+                    SELECT LATEST_FILE_HANDLES.PROJECT_ID, SUM(FILE_LATEST.CONTENT_SIZE) / POWER(1024, 3) AS TOTAL_DATA_SIZE_IN_GIB
+                    FROM LATEST_FILE_HANDLES
+                    JOIN SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILE_LATEST
+                    ON LATEST_FILE_HANDLES.FILE_HANDLE_ID = FILE_LATEST.ID
+                    GROUP BY LATEST_FILE_HANDLES.PROJECT_ID
+                )
+
+                SELECT 
+                    TOP_10_PUBLIC_PROJECTS.PROJECT_ID,
+                    TOP_10_PUBLIC_PROJECTS.N_UNIQUE_USERS,
+                    COALESCE(TO_CHAR(MAX(RECENT_DOWNLOADS.RECORD_DATE), 'YYYY-MM-DD'), 'N/A') AS LAST_DOWNLOAD_DATE,
+                    ROUND(FILE_SIZES.TOTAL_DATA_SIZE_IN_GIB, 3) AS TOTAL_DATA_SIZE_IN_GIB
+                FROM 
+                    TOP_10_PUBLIC_PROJECTS
+
+                LEFT JOIN RECENT_DOWNLOADS
+                ON TOP_10_PUBLIC_PROJECTS.PROJECT_ID = RECENT_DOWNLOADS.PROJECT_ID
+
+                JOIN FILE_SIZES
+                ON TOP_10_PUBLIC_PROJECTS.PROJECT_ID = FILE_SIZES.PROJECT_ID
+
+                GROUP BY
```

Comment:
Just some food for thought as there are many ways to solve a problem.  Another approach is to do this in two parts

1. Calculate the top 10 downloaded projects and their last download activity (`DL_INFO_CTE`)
2. Calculate the data size of those top 10 downloaded project (`DATA_SIZE_CTE`)
3. join `DL_INFO_CTE` with `DATA_SIZE_CTE` on `project_id` column.

That's effectively what you're doing here, but this `group by` clause can be a bit confusing at first glance